consensus/istanbul/backend/engine.go

```diff
@@ -406,6 +432,31 @@ func (sb *Backend) snapshot(chain consensus.ChainHeaderReader, number uint64, ha
 	}
 	sb.recents.Add(snap.Hash, snap)
 
+	validatorContract := sb.config.GetValidatorContractAddress(new(big.Int).SetUint64(targetBlockHeight))
+	if validatorContract != (common.Address{}) {
+		log.Trace("Applying snap with smart contract validators", "address", validatorContract, "client", sb.config.Client)
+
+		validatorContractCaller, err := contract.NewValidatorContractInterfaceCaller(validatorContract, sb.config.Client)
+
+		if err != nil {
+			log.Error("BFT: invalid smart contract in genesis alloc", "err", err)
+			return nil, err
```

Comment:
avoid possible double logging. can wrap error like this:
```suggestion
			return nil, fmt.Errorf("BFT: invalid smart contract in genesis alloc: %w", err)
```